File: Components/DevExpressVCL/Demos/ExpressRichEdit Control/CBuilder/Common/RibbonRichEditDemoOptions.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RibbonRichEditDemoOptions.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TRibbonDemoOptionsForm *RibbonDemoOptionsForm;

BOOL ExecuteRibbonDemoOptions(TScreenTipOptions *AScreenTipOptions, TRibbonDemoStyle *AStyle,
	TdxRibbonColorSchemeAccent *AColorSchemeAccent)
{
	TRibbonDemoOptionsForm *Form = new TRibbonDemoOptionsForm(NULL);
	Form->PopulateRibbonStyles(Form->cbRibbonStyle->Properties->Items);
	Form->PopulateColorSchemeAccents(Form->cbColorSchemeAccent->Properties->Items);
	Form->LoadOptions(*AScreenTipOptions, *AStyle, *AColorSchemeAccent);
	BOOL Result = Form->ShowModal() == mrOk;
	if (Result)
	  Form->SaveOptions(AScreenTipOptions, AStyle, AColorSchemeAccent);
	delete Form;
	return Result;
}

TdxRibbonStyle RibbonDemoStyleToRibbonStyle(TRibbonDemoStyle ADemoStyle)
{
	switch (ADemoStyle)
	{
		case rdsOffice2007:
			return rs2007;
		case rdsOffice2010:
			return rs2010;
		case rdsOffice2013:
			return rs2013;
		case rdsOffice2016:
			return rs2016;
		case rdsOffice2016Tablet:
			return rs2016Tablet;
		case rdsOffice2019:
			return rs2019;
		default: // Scenic
			return rs2010;
	}
}

//---------------------------------------------------------------------------
__fastcall TRibbonDemoOptionsForm::TRibbonDemoOptionsForm(TComponent* Owner)
		: TForm(Owner)
{
	lblColorScheme->Visible = False;
	cbColorScheme->Visible = False;
}
//---------------------------------------------------------------------------
void TRibbonDemoOptionsForm::PopulateColorSchemeAccents(TStrings *AItems)
{
	AItems->BeginUpdate();
	try
	{
		AItems->Clear();
		for (int I = (int)rcsaYellow; I <= (int)rcsaPurple; I++)
		{
			AItems->Add(dxRibbonColorSchemeAccentNames[(TdxRibbonColorSchemeAccent)I]);
		}
	}
	__finally
	{
		AItems->EndUpdate();
	}
}
//---------------------------------------------------------------------------
void TRibbonDemoOptionsForm::PopulateRibbonStyles(TStrings *AItems)
{
	AItems->BeginUpdate();
	try
	{
		AItems->Clear();
		AItems->Add("Office 2007");
		AItems->Add("Office 2010");
		AItems->Add("Office 2013");
		AItems->Add("Office 2016");
		AItems->Add("Office 2016 Tablet");
		AItems->Add("Office 2019");
		AItems->Add("Scenic");
	}
	__finally
	{
		AItems->EndUpdate();
	};
};
//---------------------------------------------------------------------------
void __fastcall TRibbonDemoOptionsForm::cbRibbonStyleSelect(TObject *Sender)
{
	lblColorSchemeAccent->Enabled = cbRibbonStyle->ItemIndex > 0;
}
//---------------------------------------------------------------------------
void TRibbonDemoOptionsForm::LoadOptions(TScreenTipOptions AScreenTipOptions, TRibbonDemoStyle AStyle,
	TdxRibbonColorSchemeAccent AColorSchemeAccent)
{
	cbRibbonStyle->ItemIndex = (Integer)AStyle;
	cbRibbonStyleSelect(NULL);

	cbColorSchemeAccent->ItemIndex = (Integer)AColorSchemeAccent;
	if (AScreenTipOptions.ShowScreenTips)
		cbScreenTipStyle->ItemIndex = (Integer)(!AScreenTipOptions.ShowDescripitons);
	else
    	cbScreenTipStyle->ItemIndex = 2;
}
//---------------------------------------------------------------------------
void TRibbonDemoOptionsForm::SaveOptions(TScreenTipOptions *AScreenTipOptions, TRibbonDemoStyle *AStyle,
	TdxRibbonColorSchemeAccent *AColorSchemeAccent)
{
	* AStyle = (TRibbonDemoStyle)cbRibbonStyle->ItemIndex;
	* AColorSchemeAccent = (TdxRibbonColorSchemeAccent)cbColorSchemeAccent->ItemIndex;
	AScreenTipOptions->ShowScreenTips = cbScreenTipStyle->ItemIndex != 2;
	AScreenTipOptions->ShowDescripitons = cbScreenTipStyle->ItemIndex == 0;	
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressRichEdit Control/Delphi/Common/RibbonRichEditDemoOptions.pas
unit RibbonRichEditDemoOptions;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, cxGraphics, cxControls, cxLookAndFeels, ExtCtrls,
  dxRibbonSkins, cxLookAndFeelPainters, cxContainer, cxEdit, cxLabel, Menus,
  cxButtons, dxBevel, cxTextEdit, cxMaskEdit, cxDropDownEdit, dxLayoutControlAdapters, dxLayoutcxEditAdapters,
  dxLayoutContainer, cxClasses, dxLayoutControl, dxForms;

type
  TRibbonDemoStyle = (rdsOffice2007, rdsOffice2010, rdsOffice2013,
    rdsOffice2016, rdsOffice2016Tablet, rdsOffice2019, rdsScenic);

  TScreenTipOptions = record
    ShowScreenTips: Boolean;
    ShowDescripitons: Boolean;
  end;

  { TRibbonDemoOptionsForm }

  TRibbonDemoOptionsForm = class(TdxForm)
    Button1: TcxButton;
    Button2: TcxButton;
    cbColorScheme: TcxComboBox;
    cbColorSchemeAccent: TcxComboBox;
    cbRibbonStyle: TcxComboBox;
    cbScreenTipStyle: TcxComboBox;
    dxLayoutControl1Group_Root: TdxLayoutGroup;
    dxLayoutControl1: TdxLayoutControl;
    dxLayoutGroup1: TdxLayoutGroup;
    dxLayoutGroup2: TdxLayoutGroup;
    dxLayoutItem1: TdxLayoutItem;
    dxLayoutItem2: TdxLayoutItem;
    dxLayoutGroup3: TdxLayoutGroup;
    liColorSchemeAccent: TdxLayoutItem;
    dxLayoutItem5: TdxLayoutItem;
    dxLayoutItem6: TdxLayoutItem;
    dxLayoutItem7: TdxLayoutItem;
    procedure cbRibbonStyleSelect(Sender: TObject);
  public
    constructor Create(AOwner: TComponent); override;

    procedure LoadOptions(const AColorSchemeName: string; const AScreenTipOptions: TScreenTipOptions;
      AStyle: TRibbonDemoStyle; AColorSchemeAccent: TdxRibbonColorSchemeAccent);
    procedure SaveOptions(out AColorSchemeName: string; out AScreenTipOptions: TScreenTipOptions;
      out AStyle: TRibbonDemoStyle; out AColorSchemeAccent: TdxRibbonColorSchemeAccent);

    class procedure PopulateColorSchemeAccents(AItems: TStrings);
    class procedure PopulateColorSchemes(AItems: TStrings; ARibbonStyle: TdxRibbonStyle);
    class procedure PopulateRibbonStyles(AItems: TStrings);
  end;

var
  RibbonDemoStyleToRibbonStyle: array[TRibbonDemoStyle] of TdxRibbonStyle = (
    rs2007, rs2010, rs2013, rs2016, rs2016Tablet, rs2019, rs2010{Scenic}
  );

function ExecuteRibbonDemoOptions(var AColorSchemeName: string; var AScreenTipOptions: TScreenTipOptions;
  var AStyle: TRibbonDemoStyle; var AColorSchemeAccent: TdxRibbonColorSchemeAccent): Boolean;
implementation

uses
  Math;

{$R *.dfm}

function ExecuteRibbonDemoOptions(var AColorSchemeName: string; var AScreenTipOptions: TScreenTipOptions;
  var AStyle: TRibbonDemoStyle; var AColorSchemeAccent: TdxRibbonColorSchemeAccent): Boolean;
begin
  with TRibbonDemoOptionsForm.Create(nil) do
  try
    LoadOptions(AColorSchemeName, AScreenTipOptions, AStyle, AColorSchemeAccent);
    Result := ShowModal = mrOk;
    if Result then
      SaveOptions(AColorSchemeName, AScreenTipOptions, AStyle, AColorSchemeAccent);
  finally
    Free;
  end;
end;

{ TRibbonDemoOptionsForm }

constructor TRibbonDemoOptionsForm.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  PopulateRibbonStyles(cbRibbonStyle.Properties.Items);
  PopulateColorSchemeAccents(cbColorSchemeAccent.Properties.Items);
end;

procedure TRibbonDemoOptionsForm.LoadOptions(const AColorSchemeName: string;
  const AScreenTipOptions: TScreenTipOptions; AStyle: TRibbonDemoStyle;
  AColorSchemeAccent: TdxRibbonColorSchemeAccent);
begin
  cbRibbonStyle.ItemIndex := Ord(AStyle);
  cbRibbonStyleSelect(nil);
  cbColorScheme.ItemIndex := Max(0, cbColorScheme.Properties.Items.IndexOf(AColorSchemeName));
  cbColorSchemeAccent.ItemIndex := Ord(AColorSchemeAccent);
  if AScreenTipOptions.ShowScreenTips then
    cbScreenTipStyle.ItemIndex := Ord(not AScreenTipOptions.ShowDescripitons)
  else
    cbScreenTipStyle.ItemIndex := 2;
end;

procedure TRibbonDemoOptionsForm.SaveOptions(out AColorSchemeName: string;
  out AScreenTipOptions: TScreenTipOptions; out AStyle: TRibbonDemoStyle;
  out AColorSchemeAccent: TdxRibbonColorSchemeAccent);
begin
  AColorSchemeName := cbColorScheme.Text;
  AStyle := TRibbonDemoStyle(cbRibbonStyle.ItemIndex);
  AScreenTipOptions.ShowScreenTips := cbScreenTipStyle.ItemIndex <> 2;
  AScreenTipOptions.ShowDescripitons := cbScreenTipStyle.ItemIndex = 0;
  AColorSchemeAccent := TdxRibbonColorSchemeAccent(cbColorSchemeAccent.ItemIndex);
end;

class procedure TRibbonDemoOptionsForm.PopulateColorSchemeAccents(AItems: TStrings);
var
  AAccentIndex: TdxRibbonColorSchemeAccent;
begin
  AItems.BeginUpdate;
  try
    AItems.Clear;
    for AAccentIndex := Low(TdxRibbonColorSchemeAccent) to High(TdxRibbonColorSchemeAccent) do
      AItems.Add(dxRibbonColorSchemeAccentNames[AAccentIndex]);
  finally
    AItems.EndUpdate;
  end;
end;

class procedure TRibbonDemoOptionsForm.PopulateColorSchemes(AItems: TStrings; ARibbonStyle: TdxRibbonStyle);

  procedure AddColorScheme(ASkin: TdxCustomRibbonSkin);
  begin
    if ASkin.Style = ARibbonStyle then
    begin
      if AItems.IndexOf(ASkin.Name) < 0 then
        AItems.Add(ASkin.Name);
    end;
  end;

var
  I: Integer;
begin
  AItems.BeginUpdate;
  try
    AItems.Clear;
    for I := 0 to dxRibbonSkinsManager.SkinCount - 1 do
      AddColorScheme(dxRibbonSkinsManager.Skins[I]);
  finally
    AItems.EndUpdate;
  end;
end;

class procedure TRibbonDemoOptionsForm.PopulateRibbonStyles(AItems: TStrings);
const
  NamesMap: array[TRibbonDemoStyle] of string = (
    'Office 2007', 'Office 2010', 'Office 2013', 'Office 2016', 'Office 2016 Tablet', 'Office 2019', 'Scenic'
  );
var
  I: TRibbonDemoStyle;
begin
  AItems.BeginUpdate;
  try
    AItems.Clear;
    for I := Low(TRibbonDemoStyle) to High(TRibbonDemoStyle) do
      AItems.Add(NamesMap[I])
  finally
    AItems.EndUpdate;
  end;
end;

procedure TRibbonDemoOptionsForm.cbRibbonStyleSelect(Sender: TObject);
var
  ASelectedColorSchemeName: string;
begin
  liColorSchemeAccent.Enabled := cbRibbonStyle.ItemIndex > 0;

  ASelectedColorSchemeName := cbColorScheme.Text;
  PopulateColorSchemes(cbColorScheme.Properties.Items, RibbonDemoStyleToRibbonStyle[TRibbonDemoStyle(cbRibbonStyle.ItemIndex)]);
  cbColorScheme.ItemIndex := Max(0, cbColorScheme.Properties.Items.IndexOf(ASelectedColorSchemeName));
end;

end.

